script/testing/util/db_server.py

```diff
@@ -227,21 +228,220 @@ def get_build_path(build_type):
+    the command used by NoisePageServer to launch the DBMS server instance specifies
+    the absolute path to the binary, relative paths (that necessarily encode the expected
+    location of the binary) will fail to function properly. Expanding these relative paths
+    to their absolute counterparts before passing them to the DBMS addresses this issue.
+
+    The above highlights the fundamental limitation of this current implementation, namely:
+
+        ALL RELATIVE PATHS ARE ASSUMED TO BE RELATIVE TO THE DBMS BINARY DIRECTORY
+
+    All relative paths that are relative to another directory will fail to resolve properly.
+
+    Arguments
+    ---------
+    value : str 
+        The DBMS server argument value
+    meta : Dict
+        Dictionary of meta-information available to all preprocessors
+
+    Returns
+    -------
+    The preprocessed server argument value
+    """
+    is_relative = str.startswith(value, "./") or str.startswith(value, "../")
+    # TODO(Kyle): This doesn't actually do any "resolving", it merely appends
+    # the absolute path to the binary directory to the relative path as it 
+    # presently exists; this works fine, but also means we might pass a needlessly
+    # long string to the DBMS server e.g. bin/././././whatever.txt is possible.
+    # It might be worth it to find some portable way to completely resolve these.
+    return os.path.join(meta["bin_dir"], value) if is_relative else value
+
+@applies_to(int, str, bool)
+def handle_flags(value, meta):
+    """
+    Handle DBMS server arguments with no associated value.
+
+    Some arguments to the DBMS are flags that do not have an associated value;
+    in these cases, we do not want to format the complete argument as
+    `-attribute=value` and instead want to format it as `-attribute` alone.
+    This preprocessor encapsulates the logic for this transformation.
+
+    TODO(Kyle): Do we actually support any arguments like this? 
+    I can't seem to come up with any actual examples...
+
+    Arguments
+    ---------
+    value : str 
+        The DBMS server argument value
+    meta : Dict
+        Dictionary of meta-information available to all preprocessors
+
+    Returns
+    -------
+    The preprocessed server argument value
+    """
+    return f"={value}" if value is not None else ""
+
+# -----------------------------------------------------------------------------
+# Utility
+
+def apply_all(functions, init_obj, meta):
+    """
+    Apply all of the functions in `functions` to object `init_obj` sequentially,
+    supplying metadata object `meta` to each function invocation.
+
+    TODO(Kyle): Initially I wanted to implement this with function composition
+    in terms of functools.reduce() which makes it really beautiful, but there
+    we run into issues with mutli-argument callbacks, and the real solution is
+    to use partial application, but this seemed like overkill... maybe revisit.
```

Comment:
comment: simple is fine and arguably better, in case someone's not a functors fan

Reply:
Agreed, definitely not a design issue to stress over, but at the time it just seemed to tantalizingly close to being a practical use of some cool functional stuff.